### PR TITLE
Add NotCheckpointSafe annotations & JPP support for generated source

### PR DESCRIPTION
**Pull request by JasonFengJ9**
Annotated `java.lang.ClassValue.initializeMap()`, `java.lang.invoke.ClassSpecializer.findSpecies()`, and `java.util.concurrent.ConcurrentHashMap.computeIfAbsent()` with `NotCheckpointSafe` to prevent `JVMCheckpointException`/deadlock;
Added `CRIU` `JPP` flag at `StandardCharsets.java.template`;
Modified `GensrcCharsetMapping.gmk` to run JPP after `StandardCharsets.java` generation.

FYI @tajila 

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

### CODE REVIEW

**Comment by JasonFengJ9** [comment]
Jenkins test sanity.functional zlinux jdk17 depends https://github.com/eclipse-openj9/openj9/pull/15593

File: make/common/JavaCompilation.gmk [hunk 1]
```diff
@@ -442,10 +468,18 @@ define SetupJavaCompilationBody
...
+		cp $(SUPPORT_OUTPUTDIR)/gensrc/java.base/sun/nio/cs/StandardCharsets.java $(SUPPORT_OUTPUTDIR)/overlay/src/java.base/sun/nio/cs
+		$$(call RunJPP, JAVA$(VERSION_FEATURE), $(SUPPORT_OUTPUTDIR)/overlay, /overlay-result)
+		cp $(SUPPORT_OUTPUTDIR)/j9jcl/overlay-result/java.base/sun/nio/cs/StandardCharsets.java $(SUPPORT_OUTPUTDIR)/gensrc/java.base/sun/nio/cs
+      endif
+
     # Create a file with all sources, to pass to javac in an @file.
     # $$($1_VARDEPS_FILE) is used as dependency to track changes in set of
     # list of files.
-    $$($1_FILELIST): $$($1_SRCS) $$($1_VARDEPS_FILE)
+    $$($1_FILELIST): $$($1_SRCS) $$($1_VARDEPS_FILE) addjpp
```
Comment (keithc-ca) [inline make/common/JavaCompilation.gmk:482]: Making all java compiles unconditional is not a reasonable solution.

I suggest we need a rule that depends on the target defined by `GensrcCharsetMapping.gmk` which runs the preprocessor on the output of that step.

File: closed/GensrcJ9JCL.gmk [hunk 2]
```diff
@@ -51,8 +51,11 @@ $(eval $(call SetupCopyFiles,COPY_OVERLAY_FILES, \
 	SRC := $(TOPDIR), \
 	DEST := $(SUPPORT_OUTPUTDIR)/overlay, \
 	FILES := \
+		src/java.base/share/classes/java/lang/ClassValue.java \
+		src/java.base/share/classes/java/lang/invoke/ClassSpecializer.java \
 		src/java.base/share/classes/java/security/Security.java \
 		src/java.base/share/classes/java/util/Timer.java \
+		src/java.base/share/classes/java/util/concurrent/ConcurrentHashMap.java \
```
Comment (keithc-ca) [inline closed/GensrcJ9JCL.gmk:58]: Please keep packages grouped together (`java/util/concurrent` should go after the last entry for `java/util`).

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: make/common/JavaCompilation.gmk (hunk 1 above)
Comment (JasonFengJ9) [reply make/common/JavaCompilation.gmk]: Moved JPP call into `GensrcCharsetMapping.gmk`.

File: closed/GensrcJ9JCL.gmk (hunk 2 above)
Comment (JasonFengJ9) [reply closed/GensrcJ9JCL.gmk]: Updated.

**Comment by JasonFengJ9** [comment]
Jenkins test sanity.functional xlinux jdk17

File: make/modules/java.base/gensrc/GensrcCharsetMapping.gmk [hunk 3]
```diff
@@ -52,5 +76,10 @@ $(CHARSET_DONE_BASE)-stdcs: $(CHARSET_DATA_DIR)/charsets \
 	    $(CHARSET_COPYRIGHT_HEADER) \
             $(LOG_DEBUG)
 	$(TOUCH) '$@'
+	$(ECHO) OpenJ9Logs0725-1428pm - '$@'
+	$(MKDIR) -p $(SUPPORT_OUTPUTDIR)/overlay-gensrc/src/java.base/sun/nio/cs
+	$(CP) $(CHARSET_GENSRC_JAVA_DIR_BASE)/StandardCharsets.java $(SUPPORT_OUTPUTDIR)/overlay-gensrc/src/java.base/sun/nio/cs
+	$(call RunJPP, JAVA$(VERSION_FEATURE), $(SUPPORT_OUTPUTDIR)/overlay-gensrc, /overlay-result)
+	$(CP) $(SUPPORT_OUTPUTDIR)/j9jcl/overlay-result/java.base/sun/nio/cs/StandardCharsets.java $(CHARSET_GENSRC_JAVA_DIR_BASE)
```
Comment (keithc-ca) [inline make/modules/java.base/gensrc/GensrcCharsetMapping.gmk:83]: This should go before the `$(TOUCH)` command above so failures won't be ignored on retry.

Comment (keithc-ca) [reply make/modules/java.base/gensrc/GensrcCharsetMapping.gmk:83]: JPP output should go somewhere other than `$(J9JCL_SOURCES_DIR)/overlay-result` lest it be confused for a module.

**Review by keithc-ca — CHANGES_REQUESTED** [review]
Please remove your changes to `JavaCompilation.gmk` from both commits (or squash).

File: src/java.base/share/classes/java/lang/ClassValue.java [hunk 4]
```diff
@@ -34,6 +40,10 @@
 import static java.lang.ClassValue.ClassValueMap.probeHomeLocation;
 import static java.lang.ClassValue.ClassValueMap.probeBackupLocations;
 
+/*[IF CRIU_SUPPORT]*/
+import openj9.internal.criu.NotCheckpointSafe;
+/*[ENDIF] CRIU_SUPPORT*/
```
[the remark this answers is not in the document]
Comment (JasonFengJ9) [reply src/java.base/share/classes/java/lang/ClassValue.java]: Added spaces.

File: make/modules/java.base/gensrc/GensrcCharsetMapping.gmk (hunk 3 above)
Comment (JasonFengJ9) [reply make/modules/java.base/gensrc/GensrcCharsetMapping.gmk]: Changed the output folder to `$(SUPPORT_OUTPUTDIR)/overlay-result`.

**Comment by JasonFengJ9** [comment]
Squashed the commits, obsoleted https://github.com/ibmruntimes/openj9-openjdk-jdk17/pull/127.

**Comment by JasonFengJ9** [comment]
Jenkins test sanity.functional zlinux jdk17

File: make/modules/java.base/gensrc/GensrcCharsetMapping.gmk [hunk 5]
```diff
@@ -51,6 +75,10 @@ $(CHARSET_DONE_BASE)-stdcs: $(CHARSET_DATA_DIR)/charsets \
 	    $(CHARSET_STANDARD_JAVA_TEMPLATES) $(CHARSET_EXTSRC_DIR) \
 	    $(CHARSET_COPYRIGHT_HEADER) \
             $(LOG_DEBUG)
+	$(MKDIR) -p $(SUPPORT_OUTPUTDIR)/overlay-gensrc/src/java.base/sun/nio/cs
+	$(CP) $(CHARSET_GENSRC_JAVA_DIR_BASE)/StandardCharsets.java $(SUPPORT_OUTPUTDIR)/overlay-gensrc/src/java.base/sun/nio/cs
+	$(call RunJPP, JAVA$(VERSION_FEATURE), $(SUPPORT_OUTPUTDIR)/overlay-gensrc, /overlay-result)
+	$(CP) $(SUPPORT_OUTPUTDIR)/overlay-result/java.base/sun/nio/cs/StandardCharsets.java $(CHARSET_GENSRC_JAVA_DIR_BASE)
```
Comment (keithc-ca) [inline make/modules/java.base/gensrc/GensrcCharsetMapping.gmk:81]: For clarity, I think these two `$(CP)` commands should explicitly use a trailing slash to indicate the target is a directory:
```
	$(CP) $(CHARSET_GENSRC_JAVA_DIR_BASE)/StandardCharsets.java $(SUPPORT_OUTPUTDIR)/overlay-gensrc/src/java.base/sun/nio/cs/
	$(call RunJPP, JAVA$(VERSION_FEATURE), $(SUPPORT_OUTPUTDIR)/overlay-gensrc, /overlay-result)
	$(CP) $(SUPPORT_OUTPUTDIR)/overlay-result/java.base/sun/nio/cs/StandardCharsets.java $(CHARSET_GENSRC_JAVA_DIR_BASE)/
```

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: make/modules/java.base/gensrc/GensrcCharsetMapping.gmk (hunk 5 above)
Comment (JasonFengJ9) [reply make/modules/java.base/gensrc/GensrcCharsetMapping.gmk]: Appended `/`.

**Review by keithc-ca — APPROVED** [review]
The latest changes don't invalidate the successful testing in https://openj9-jenkins.osuosl.org/view/Pull%20Requests/job/PullRequest-OpenJDK17/68.

**Pull request merged**